### DISCUSSION TITLE
PID control

### DISCUSSION THREAD

**Discussion by Harvie** [vmatare/thinkfan#216] [Ideas]
I think it shouldn't be hard to implement PID control and it might even vastly simplify the configuration and stabilize the RPMs even more.

**Comment by vmatare** [comment]
It's certainly not hard to implement. In fact I tried it once, but I came to the conclusion that the ID part wasn't actually useful. Understand that PID has a different goal, which is to adjust some target measure as precisely and quickly as possible, with as little oscillation as possible. For fan control, we need neither the precision nor the quickness. There simply isn't some single target temperature that we want keep at all costs. So to get any remotely practical fan behavior, PID is effectively reduced to P, which then comes down to simple interpolation between given (Temp,Speed) tuples. That is something I'm actually considering.

**Comment by Harvie** [comment]
> For fan control, we need neither the precision nor the quickness. There simply isn't some single target temperature that we want keep at all costs.

OK. Good point.

> PID is effectively reduced to P, which then comes down to simple interpolation between given (Temp,Speed) tuples. That is something I'm actually considering.

Yes. That might simplify the configuration.

**Comment by vmatare** [comment]
I've given this some more thought, and what might be worth a shot is a modification of PID, where the target measure isn't a point, but an interval. The difficulty with that is that a sharp cutoff at the interval's borders would defeat the purpose of the differential component...

**Comment by Harvie** [comment]
I think the D part is important... Otherwise fan noise might oscillate. In my experience, it can happen that fan reacts too slowly to the temperature change, which means system heats up and suddenly fan has to go all the way up, which cools everything and fan shuts down for a while... such oscillation can repeat itself forever... I prevent this by never going to the full stop, cooling just little bit more than is actualy needed. My target here is to have stable cooling performance with little to no noise.
